### ISSUE TITLE
[Bug] Compiling fails when server starts

### ISSUE THREAD

**Issue by alvinkl** [webpack/webpack-dev-middleware#347]
* Operating System:
* Node Version: v10.11.0
* NPM Version: 6.4.1
* webpack version: 4.20.2
* webpack-dev-middleware Version: 3.4.0

- [ ] This is a **feature** request
- [X] This is a **bug**

### Background
I am building a framework and using webpack-dev-middleware for HMR on development and running it on example project but webpack-dev-middleware doesn't work as expected.

### Code
/build/webpack/webpack.ts
```js
...
 /**
   * Setting Up Default Rules
   */
  const getDefaultRules = (): RuleSetRule[] => {
    return [
      {
        test: /\.jsx?/,
        exclude: /node_modules/,
        use: 'zones-babel-loader',
      },
    ];
  };
...
let baseConfigs: Configuration = {
    name,
    entry: getDefaultEntry(),
    mode,
    devtool: isDev || sourceMaps ? 'cheap-module-source-map' : false,
    cache,
    bail: true,
    performance: { hints: false },
    resolve: {
      extensions: ['*', '.js', '.jsx', '.json'],
      alias,
      modules: [
        ZONES_PROJECT_ROOT_NODE_MODULES,
        'node_modules',
        ...nodePathList, // Support for NODE_PATH environment variable
      ],
    },
    resolveLoader: {
      symlinks: true,
      modules: [
        ZONES_PROJECT_ROOT_NODE_MODULES,
        'node_modules',
        path.join(__dirname, '..', 'loaders'), // The loaders Zones provides
        ...nodePathList, // Support for NODE_PATH environment variable
      ],
    },
    optimization: getDefaultOptimizations(),
    target: isServer ? 'node' : 'web',
    module: {
      rules: getDefaultRules(),
    },
    plugins: getDefaultPlugins(),
    output: getDefaultOutput(),
  };

  if (isServer) {
    baseConfigs = {
      ...baseConfigs,

      // Extract all server node_modules from bundles except React
      externals: [
        nodeExternals({
          whitelist: [/react/],
        }),
      ],

      node: {
        // Keep __dirname relative to file not root project
        __dirname: true,
      },
    };
  }
....
```

The custom loader
/build/zones-babel-loader.ts
```js
import babelLoader from 'babel-loader';
import { preset } from '../babel/preset';

export default babelLoader.custom((babel) => {
  const presetConfig = babel.createConfigItem(preset, { type: 'preset' });

  const configs = new Set();

  return {
    config(config) {
      const options = { ...config.options };
      if (config.hasFilesystemConfig()) {
        for (const file of [config.babelrc, config.config]) {
          if (file && !configs.has(file)) {
            configs.add(file);
            console.log('[Zones-Babel-Loader] Using external babel configuration');
            console.log('[Zones-Babel-Loader] File: ', file);
          }
        }
      } else {
        options.presets = [...options.presets, presetConfig];
      }

      return options;
    },
  };
});
```

/build/babel/preset.ts
```js
const env = process.env.NODE_ENV;
const isDevelopment = env === 'development';
const isTest = env === 'test';

export const preset = (context, opts = {}) => {
  const presets = [
    [
      require('@babel/preset-env').default,
      {
        modules: false,
        // cacheDirectory
        ...opts['preset-env'],
      },
    ],
    [
      require('@babel/preset-react'),
      {
        development: isDevelopment || isTest,
        ...opts['preset-react'],
      },
    ],
  ];

  const plugins = [
    [
      require('@babel/plugin-proposal-class-properties'),
      opts['class-properties'] || {
        loose: true,
      },
    ],
    [
      require('@babel/plugin-proposal-decorators'),
      opts['proposal-decorators'] || {
        legacy: true,
      },
    ],
    require('@babel/plugin-proposal-object-rest-spread'),
    require('@babel/plugin-syntax-dynamic-import'),
    require('react-loadable/babel'),
    require('./plugins/react-loadable-plugin'),
    // for development
    [
      require('@babel/plugin-transform-runtime'),
      {
        regenerator: true,
        helpers: false,
        corejs: 2,
        useESModules: false,
      },
    ],
    // for production
    // require('@babel/runtime'),
  ].filter(Boolean);

  return {
    presets,
    plugins,
  };
};
```

Server Code
```js
...
class Server {
  ...
  private setupDevServer = (app: Express) => {
    const configs = getConfigs({ env: 'development' });
    const compiler = webpack(configs);

    const wdmOptions = {
      stats: false,
      serverSideRender: true,
    };
    app.use(require('webpack-dev-middleware')(compiler, wdmOptions));

    app.use(require('webpack-hot-middleware')(compiler));
  };
  ...
}
```

package.json
```json
{
    "name": "@zones/framework",
    "version": "1.0.0",
    "description": "",
    "main": "index.js",
    "scripts": {
        "build": "rm -rf dist && gulp",
        "watch": "gulp watch"
    },
    "author": "",
    "license": "ISC",
    "private": false,
    "sideEffects": false,
    "dependencies": {
        "@babel/core": "7.1.2",
        "@babel/plugin-proposal-class-properties": "7.1.0",
        "@babel/plugin-proposal-decorators": "7.1.2",
        "@babel/plugin-proposal-object-rest-spread": "7.0.0",
        "@babel/plugin-syntax-dynamic-import": "7.0.0",
        "@babel/plugin-transform-runtime": "7.1.0",
        "@babel/preset-env": "7.1.0",
        "@babel/preset-react": "7.0.0",
        "@babel/runtime": "7.1.2",
        "@babel/runtime-corejs2": "7.1.2",
        "autodll-webpack-plugin": "^0.4.2",
        "babel-loader": "8.0.4",
        "es6-promise": "4.2.5",
        "helmet": "3.14.0",
        "isomorphic-fetch": "2.2.1",
        "node-notifier": "5.2.1",
        "react-helmet": "5.2.0",
        "react-router-config": "^4.4.0-beta.6",
        "tapable": "1.1.0",
        "uglifyjs-webpack-plugin": "2.0.1",
        "webpack": "4.20.2",
        "webpack-cli": "3.1.2",
        "webpack-dev-middleware": "3.4.0",
        "webpack-hot-middleware": "2.24.3",
        "webpack-node-externals": "^1.7.2",
        "webpackbar": "2.6.3"
    },
    "peerDependencies": {
        "express": "4.16.4",
        "react": "16.5.2",
        "react-dom": "16.5.2",
        "react-loadable": "5.5.0",
        "react-router": "^4.4.0-beta.1",
        "react-router-dom": "^4.4.0-beta.1"
    },
    "devDependencies": {
        "@babel/preset-typescript": "^7.1.0",
        "@types/webpack-bundle-analyzer": "^2.13.0",
        "babel-plugin-module-resolver": "^3.1.1",
        "express": "4.16.4",
        "gulp": "gulpjs/gulp.git#4.0",
        "gulp-babel": "^8.0.0",
        "gulp-clean": "^0.4.0",
        "gulp-sourcemaps": "^2.6.4",
        "gulp-uglify": "^3.0.1",
        "gulp-watch": "^5.0.1",
        "react": "16.5.2",
        "react-dom": "16.5.2",
        "react-loadable": "5.5.0",
        "react-router": "^4.4.0-beta.1",
        "react-router-dom": "4.3.1",
        "webpack-bundle-analyzer": "^3.0.3",
        "webpack-stats-plugin": "^0.2.1"
    }
}

```

### Expected Behavior
webpack-dev-middleware to run properly
### Actual Behavior
Webpack does compile successfully when I build the project, but when I start the server it throws an error:

Terminal
```sh
Server running on port :8080
(node:23425) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'default' of undefined
    at /home/alvinkristanto/github/project/app-infra/zones/example/.zones/server.js:209901:67
    at process._tickCallback (internal/process/next_tick.js✔ success server compiled in 72ms
Type checking in progress...
(node:23425) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'default' of undefined
    at /home/alvinkristanto/github/project/app-infra/zones/example/.zones/server.js:209901:67
    at process._tickCallback (internal/process/next_tick.js:68:7)
(node:23425) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
✔ success client compiled in 277ms
```
and on browser:
```
Module build failed (from ../packages/zones/dist/build/loaders/zones-babel-loader.js):
Error: Cannot find module '/home/alvinkristanto/github/project/app-infra/zones/packages/zones/dist/build/loaders/zones-babel-loader.js'
    at /home/alvinkristanto/github/project/app-infra/zones/example/.zones/server.js:233795:11
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

I used VSCode debugger and found out the error comes from here 
```js
/***/ "pwHm":
/*!*******************************************************!*\
  !*** ../node_modules/loader-runner/lib/loadLoader.js ***!
  \*******************************************************/
/*! no static exports found */
/***/ (function(module, exports, __webpack_require__) {

module.exports = function loadLoader(loader, callback) {
	if(true) {
		__webpack_require__("u1yz")(loader.path).catch(callback).then(function(module) {
			loader.normal = typeof module === "function" ? module : module.default;
			loader.pitch = module.pitch;
			loader.raw = module.raw;
			if(typeof loader.normal !== "function" && typeof loader.pitch !== "function")
				throw new Error("Module '" + loader.path + "' is not a loader (must have normal or pitch function)");
			callback();
		});
	} else { var retry, module; }
};
```

loader.path refers to `/home/alvinkristanto/github/project/app-infra/zones/packages/zones/dist/build/loaders/zones-babel-loader.js`

If I run `$ cat /home/alvinkristanto/github/project/app-infra/zones/packages/zones/dist/build/loaders/zones-babel-loader.js`,
```js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/toConsumableArray"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/objectSpread"));

var _set = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/set"));

var _babelLoader = _interopRequireDefault(require("babel-loader"));

var _preset = require("../babel/preset");

var _default = _babelLoader.default.custom(function (babel) {
  var presetConfig = babel.createConfigItem(_preset.preset, {
    type: 'preset'
  });
  var configs = new _set.default();
  return {
    config: function config(_config) {
      var options = (0, _objectSpread2.default)({}, _config.options);

      if (_config.hasFilesystemConfig()) {
        var _arr = [_config.babelrc, _config.config];

        for (var _i = 0; _i < _arr.length; _i++) {
          var file = _arr[_i];

          if (file && !configs.has(file)) {
            configs.add(file);
            console.log('[Zones-Babel-Loader] Using external babel configuration');
            console.log('[Zones-Babel-Loader] File: ', file);
          }
        }
      } else {
        options.presets = (0, _toConsumableArray2.default)(options.presets).concat([presetConfig]);
      }

      return options;
    }
  };
});

exports.default = _default;
//# sourceMappingURL=zones-babel-loader.js.map
```

**Comment by evilebottnawi** [comment]
@alvinkl Please create minimum reproducible test repo

**Comment by evilebottnawi** [comment]
Also looks problem in loader

**Comment by alvinkl** [comment]
Here you go https://github.com/alvinkl/wdm-example-error

To reproduce the error, you can try to run
```sh
$ yarn
$ yarn build # to build the framework 
$ yarn example:build # build the example project using the framework
$ yarn example:run
```

The Webpack config can be found at `packages/framework/build/webpack`
The custom loader I made can be found at `packages/framework/build/loaders`
The default server that execute WDM can be found at `packages/framework/server/

**Comment by alvinkl** [comment]
If the problem is in the loader, I wonder why it builds successfully before I apply WDM, am I missing something here?

**Comment by alvinkl** [comment]
@evilebottnawi here you go https://github.com/alvinkl/wdm-example-error. I've reduced the repo code.

**Comment by evilebottnawi** [comment]
@alvinkl thanks!

**Comment by evilebottnawi** [comment]
@alvinkl Why you build webpack? Open you `.zoners/server.js` and look on file content, you build `webpack` using `webpack` with all plugins and loaders and many of their doesn't work in browsers. It is not problem in `webpack-dev-middleware`. You can find here information how use can use `webpack` with existing server https://github.com/webpack-contrib/webpack-hot-middleware. Also when you need create minimum reproducible test repo, please avoid using `lerna`, minimum `ts` because not all people use `ts`.

In short double check all lines in your configuration. I don't think you need `webpack` in bundle.

**Comment by alvinkl** [comment]
@evilebottnawi oh wow!!! that was a silly mistake, I build the `webpack` with all the plugins inside it.
I was debugging this for more than 2 days and kinda frustrated a bit. All because that silly `Server` config :disappointed: 
Thank you so much for your help.
I think we can close this issue since this is not a bug from WDM but rather my code.
